### PR TITLE
Roberto - For Dev Branch- Adds email notification for Owner create/reset password

### PR DESCRIPTION
**Pull request by robertoooc**
# Description
Screenshot 2023-10-10 at 1 34 00 PM
This PR will send an email when a new Owner account is either created or their password is reset.

## Related PRS (if any):
Isn't related to any PR but test with latest frontend

## Main changes explained:
- Imports and uses emailSender function in userProfileController.js. Email is sent when a new owner account is created or when a owner's password is reset. 
- Adds more security checks to the reset password function because it didn't originally check permissions.

# IMPORTANT SET UP
1. check into current branch
2. You need to create your own OAuth credentials - Watch this video up to 6:56 ([Where to get new env variables](https://www.youtube.com/watch?v=-rcRf7yswfM&t=10s)) 
3. If the video above gives you issues at 6:00 when the app isn't verified and doesn't let you proceed, Watch this video to fix that ([How to fix the OAuth Error](https://www.youtube.com/watch?v=x_NPvk0pk8g))
4. Add these variables to your environment file with your own credentials
`REACT_APP_EMAIL=The email you use for the OAuth`
`REACT_APP_EMAIL_CLIENT_ID=Watch the video to get this`
`REACT_APP_EMAIL_CLIENT_SECRET=Watch the video to get this`
`REACT_APP_EMAIL_REDIRECT_URI=https://developers.google.com/oauthplayground`
`REACT_APP_EMAIL_REFRESH_TOKEN=Watch the video to get this`
`sendEmail=true`
5. In `src/controllers/userProfileController.js`, change the emails to whatever recipient email you wish to use on lines `249` and `923` (ex: onecommunityglobal@gmail.com -> yourEmail@gmail.com) -It should be different from the email in the .env file
6. Do `npm install`, `npm run build`, and `npm start`  (Make sure to run build after making any changes)
7. In the frontend check into development and log into an Owner account
8. Verify that creating an Owner account results in an email sent to your recipient
9. Verify that resetting an Owner account's password results in an email sent to your recipient

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/117053202/75d5eeb7-386c-4749-a681-34929eefbcd8

## Note:
- If you have any issues in the how to test section check out these docs or feel free to ping me for help: 
[Implement a feature related to Email sending and testing](https://docs.google.com/document/d/1HJJTDtWykjGb4IfiIFlHXDfp_IABvmwKxsMPqqF6rG8/edit)
[Update Oauth 2.0 Gmail Email Token Refresh Instructions](https://docs.google.com/document/d/1By_Ls6RGbesMgZvRzU2zIPIJtLBKMgQjabTUH2ZM8kE/edit)

- To prevent unintended email notifications, it is crucial to stop the local server running on your machine before midnight Pacific Time on Saturday. This precaution ensures that the cron job does not execute on your local server and random emails are not sent to everyone.

### CODE REVIEW

**Review by StrawberryCalpico — COMMENTED** [review]
Screenshot 2023-10-17 at 5 35 56 PM
Screenshot 2023-10-17 at 5 36 24 PM



Hi, I tried to tested your PR. I followed the instructions and set up the google API, but I didn't received the email after creating new "owner".  I kept receiving error messages(400). I added a console.log in the try catch [emailSender.js](https://github.com/OneCommunityGlobal/HGNRest/blob/728b83d1dcf53b48e248b5864e10449afa20b5ec/src/utilities/emailSender.js#L64C34-L64C34) to print the error message.

**Review by malikjahanzaib — APPROVED** [review]
Hi @robertoooc, I tested this PR and everything works great. When an owner account is created or there is a password reset for an owner account, an email is sent out. Great job!!!

Screenshots:

Screenshot 2023-10-21 033151

Screenshot 2023-10-21 033201

Screenshot 2023-10-21 033034

**Review by shubhankarval — APPROVED** [review]
Hey @robertoooc, I tested this PR and it works exactly as expected, good job! 
\
Creating Owner account:
pr
pr2
\
Resetting password for Owner account:
pr3
pr4

File: src/controllers/userProfileController.js [hunk 1]
```diff
@@ -826,28 +859,78 @@ const userProfileController = function (UserProfile) {
...
-          })
-          .catch((error) => {
-            res.status(500).send(error);
-          });
-      })
-      .catch((error) => {
-        res.status(500).send(error);
+      ValidatePassword(req);
+
+      const requestor = await UserProfile.findById(req.body.requestor.requestorId).select('firstName lastName email role').exec();
```
Comment (John-Flavian) [inline src/controllers/userProfileController.js:867]: Nicely done.

Although good practice of "seperation of concerns" would be nice.

**Review by Shuhua-L — APPROVED** [review]
Hi @robertoooc, I successfully tested the PR, and everything is working as intended. Nicely done!

To be more specific:
- I logged in as an Owner (Shuhua Owner), created a new owner account (Shuhua Owner1), and received an email
  PR569-1
  PR569-2

- I then reset the password of the new account and received another notification
  PR569-3

**Comment by ilyaflaks** [comment]
@robertoooc I tested this PR. The "reset password" functionality works just as intended! Good job!
969-password-reset-email

However, when I try to create a new user, I get a 502 Server error: 
969branch-front-end-error


When I console log the error on the backend, the message is: **userProfile validation failed: location: Cast to string failed for value "{...}" (type Object) at path "location"**
969branch-save-error

When I console log **req.body.location**, it looks like this: 
969branch-create-error


I can't immediately see how the changes you made to your code can lead to this error. When I create a user in the "development" branch with the same information I don't get any errors. The object for req.body.location looks the same: 
dev-branch-create-success

This is also the case for me for the branch **Roberto_Create_Notification_For_Main_Branch**

**Review by XiaohanMeng — APPROVED** [review]
I have tested this Pull Request with my owner account, and it worked as intended. 
First, I logged into my owner account and created a test account with the owner role. Here is the email I received. 
PR569_result1


Then, I reset the password for the test owner account and received a new email.
PR569_result2

**Review by huangweihou — APPROVED** [review]
Hi @robertoooc, I tested an owner account was created and a password reset for an owner account. Both worked correctly with emails received. Great job!!!
Please take below screenshots as reference:
Screen Shot 2024-02-17 at 11 28 59 PM
Screen Shot 2024-02-17 at 11 28 59 PM

**Comment by psharma1984** [comment]
Hi @robertoooc, tested with an owner account. As a new user is created and a password is reset for an owner account, it sends an email to the recipient.
Please take below screenshots as reference:
Screenshot from 2024-02-19 17-53-55
Screenshot from 2024-02-19 17-49-08

**Comment by psharma1984** [comment]
Hi @robertoooc, I tested an owner account was created and a password reset for an owner account. Both worked correctly with emails received. Great job!!!
Please take below screenshots as reference:
Screenshot from 2024-02-19 17-53-55
Screenshot from 2024-02-19 17-49-08

**Review by psharma1984 — APPROVED** [review]

**Review by skaush21 — APPROVED** [review]
features work well!
* followed instructions on video -> followed instructions on second video as well
* added specified code to .env
* navigated to src/controllers/userProfileController.js added new email to lines 274 and 946.
* npm install, npm run build, npm run start, did the same for frontend
* logged in using owner account
* created owner account, received email verification
* reset owner account password, received email verification
All mentioned features worked well, great job!
I do want to point out that there is a console warning, although I dont think its anything to worry about. 
account appears on user list
create account
reset password
Screen Shot 2024-02-26 at 10 27 40 PM

**Review by ramyaram2092 — CHANGES_REQUESTED** [review]
Hello @robertoooc  Good job on the functionality! .I did get an email for account creation. However I didnt recieve any for password update 
image

**Comment by robertoooc** [comment]
@ramyaram2092, It looks like the problem was at some point the allusers cache was not defined or didn't return anything so the previous code that would try to add a user or remove one from that array would cause that error, added an if statement as error handling to prevent that case now.

Also before sending email checking the dbName if its the hgn dev then it only alerts for Owner creation and password reset, but if the dbName is anything else such as the main site's db then it would also alert of Admin and Owner actions.

**Review by ramyaram2092 — COMMENTED** [review]
Check the following review

**Review by ramyaram2092 — APPROVED** [review]
Works @robertoooc . Good job !. I got email for both scenarios

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**